kedro/io/data_catalog.py

```diff
@@ -170,7 +190,14 @@ def __init__(
         self._data_sets = dict(data_sets or {})
         self.datasets = _FrozenDatasets(self._data_sets)
         self.layers = layers
-
+        # Keep a record of all patterns in the catalog.
+        # {dataset pattern name : dataset pattern body}
+        self.dataset_patterns = dict(dataset_patterns or {})
+        self._sorted_dataset_patterns = sorted(
+            self.dataset_patterns.keys(),
+            key=lambda x: (_specificity(x), -x.count("{"), x),
+        )
+        self._pattern_name_matches_cache: dict[str, str] = {}
```

Comment:
I don't quite understand what this is for. If the dataset has been matched it's added to the catalog and will be in `self._data_sets` so why exactly do you need this?